File: app/src/main/java/com/example/mnowak/cirriculumvitae/ui/fragments/ExperienceFragment.java
package com.example.mnowak.cirriculumvitae.ui.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.mnowak.cirriculumvitae.R;
import com.example.mnowak.cirriculumvitae.models.CompanyViewModel;
import com.example.mnowak.cirriculumvitae.ui.adapters.CompaniesRecyclerViewAdapter;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ExperienceFragment extends Fragment {
    private static final String ARG_EXPERIENCE_MODEL = "experience_model";

    private List<CompanyViewModel> experience;

    @BindView(R.id.rvCompanies)
    RecyclerView rvCompanies;

    public static ExperienceFragment newInstance(List<CompanyViewModel> experience) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_EXPERIENCE_MODEL, experience.toArray(new CompanyViewModel[]{}));
        ExperienceFragment fragment = new ExperienceFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        Log.e("EXP", "CREATE");
        super.onCreate(savedInstanceState);
        readArguments();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        Log.e("EXP", "CREATE VIEW");

        return inflater.inflate(R.layout.fragment_experience, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Log.e("EXP", "CREATEd");

        ButterKnife.bind(this, view);
        setupRecyclerView();
    }

    private void readArguments() {
        if (getArguments() != null) {
            Serializable serializableExperience = getArguments().getSerializable(ARG_EXPERIENCE_MODEL);
            if (serializableExperience != null && serializableExperience instanceof CompanyViewModel[]) {
                experience = Arrays.asList((CompanyViewModel[]) serializableExperience);
            }
        }
    }

    private void setupRecyclerView() {
        rvCompanies.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        rvCompanies.setAdapter(new CompaniesRecyclerViewAdapter(experience, getContext()));
    }
}

File: app/src/main/java/com/mnowak/curriculumvitae/feature/moreInfo/fragment/experience/ExperienceFragment.kt
package com.mnowak.curriculumvitae.feature.moreInfo.fragment.experience

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.mnowak.curriculumvitae.R
import com.mnowak.curriculumvitae.databinding.FragmentExperienceBinding
import com.mnowak.curriculumvitae.di.viewModel.ViewModelFactory
import com.mnowak.curriculumvitae.feature.moreInfo.fragment.experience.list.ExperienceAdapter
import dagger.android.support.DaggerFragment
import javax.inject.Inject

class ExperienceFragment : DaggerFragment() {

    @Inject
    lateinit var viewModelFactory: ViewModelFactory
    lateinit var binding: FragmentExperienceBinding

    private val adapter = ExperienceAdapter(this)

    private val viewModel: ExperienceViewModel by lazy {
        ViewModelProviders.of(this, viewModelFactory).get(ExperienceViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentExperienceBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupRecyclerView()
        observe()
    }

    private fun setupRecyclerView() {
        binding.rvCompanies.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        binding.rvCompanies.adapter = adapter
    }

    private fun observe() {
        viewModel.experience.observe(viewLifecycleOwner, Observer {
            adapter.submitList(it)
        })
    }

    companion object {

        fun newInstance(): ExperienceFragment {
            return ExperienceFragment()
        }
    }
}
